Proposal for improving addons

SimonTheLeg opened the pull request:
Signed-off-by: Simon Bein <simontheleg@gmail.com>

**What this PR does / why we need it**:

Proposal for Improving Addons. Why we need it → see Motivation and Background Section + #8047 

**Does this PR introduce a user-facing change?**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If  no release note is required, just write "NONE".
-->
```release-note
NONE
```

SimonTheLeg committed:
Proposal for improving addons

Signed-off-by: Simon Bein <simontheleg@gmail.com>

embik commented on docs/proposals/improve-addons.md:
```diff
@@ -0,0 +1,350 @@
+- CoreAddonCatalogue → will be maintained by Kubermatic. Should not be directly edited by the customer. We think it makes sense to add this as a layer of protection against misconfiguration. This catalogue is rarely changed. Moreover, this catalogue will not be shown in the UI. This catalogue contains addons for the supported Kubernetes versions (eg KubeProxy will have 3 versions of the addons, o ...
+- CustomAddonCatalogue(s) →one or multiple catalogues that contain custom addons. This catalogue can be changed more often. Additionally, it is possible to have multiple CustomAddonCatalogues
+
+Furthermore, the AddonCatalogueDistributor ensures that there are no duplicate addons inside the AddonCatalogue. A duplicate is defined by two addons having the same `name` and `version`. In these situations, we propose the AddonCatalogueDistributor to stop its distribution and log an error.
```
I'd suggest to validate the `AddonCatalogue` resources through a webhook instead, so it gets rejected and doesn't take down the AddonCatalogueDistributor as a whole.

embik replied:
Not sure we're talking about the same thing here :D I was referring to the duplicate detection in an (instance of) `AddonCatalogue`. Maybe I misunderstood this part of the proposal, but it looked like the validation of resources was supposed to happen in the distributor, while a webhook would help by rejecting invalid `AddonCatalogues` before they hit the cluster and can be processed by the AddonCatalogueDistributor.

SimonTheLeg replied:
I commented on the wrong comment on accident. I have moved my comment since then. Will get back to you about the webhook soon. Just have to think about it for a minute

SimonTheLeg replied:
Good point with the WebHook, we have added it to the proposal.

embik commented on docs/proposals/improve-addons.md:
```diff
@@ -0,0 +1,350 @@
+spec:
+  coreAddon: false #This can be omitted, false will be the default
+    targetNamespace: monitoring
+    createNamespace: true # flag to create the targetNamespace
```
this in't valid YAML. I'm not sure if it's intentional or not, but it's also happening in the example under `CoreAddonController`, so I wanted to point it out. Probably looking for:

```suggestion
spec:
  coreAddon: false #This can be omitted, false will be the default
  targetNamespace: monitoring
  createNamespace: true # flag to create the targetNamespace
```

SimonTheLeg replied:
Good Spot! 
Yeah sometimes Nuclino Yaml hates you :D

embik commented on docs/proposals/improve-addons.md:
```diff
@@ -0,0 +1,350 @@
+## Goals
+
+- Decouple custom addons from the KKP lifecycle
+- Handle dependency **inside** an Addon (e.g. CRD need to be created before CRs, namespace needs to be created before other objects)
```
How is this goal achieved in the proposal? Maybe I missed it, but I couldn't find it. The only thing I found is `RequiredResourceTypes`, and my understanding is that this is actually defining outside dependencies (prerequisites for successfully installing the addon).

SimonTheLeg replied:
Ah interesting. Maybe I misunderstood our current implementation. I thought our current implementation would do exactly that: ensure that CRDs and namespaces are deployed before any other manifests.
cc @vgramer: This is something I think we have to look in the code of the current implementation. But yes, if the current system is not able to install CRDs before deploying resources, then we should include that capability in our proposal.

On the other side: For the helm rendering method, handling dependencies is done directly via helm; So theoretically this should come out of the box (of course during implementation we will have to see how well it works in practice)

embik replied:
Ah, I was more thinking of a flexible dependency management system within the addon, like https://carvel.dev/kapp/docs/latest/apply-ordering/. That's probably fairly complex to pull off, but I misunderstood the goal then :)

vgramer replied:
@SimonTheLeg yes the `RequiredResourceTypes` works as @embik explain. it's a pre-requisite to render that addon's manifests.

Concerning the `Handle dependency **inside** an Addon ` I think too that `helm` should do the job

SimonTheLeg replied:
@embik we have updated the proposal to reflect that a change to the go-template method will be made. You can find the addendum in the `AddonInstallationController -> go-template` section of this document

embik commented on docs/proposals/improve-addons.md:
```diff
@@ -0,0 +1,350 @@
+  - template
+    - a rendering method. Possible methods are (additional methods can easily be added in the future):
+      - go-template → same as what we currently offer in addons
+      - helm → rendering of helm charts
```
Any plans to make the template data (see https://docs.kubermatic.com/kubermatic/v2.18/guides/addons/#manifest-templating; this could help generating per-cluster Ingress FQDNs or other resources based on cluster information) passed to the current addons (`go-template` rendering method in this proposal) accessible to helm-based addons?

SimonTheLeg replied:
that is actually a really good point. We would have to slightly adapt the proposal, but this makes sense to me

SimonTheLeg replied:
we have added it into the proposal

embik reviewed (commented):
I've added a few (minor) comments.

maciaszczykm commented on docs/proposals/improve-addons.md:
```diff
@@ -0,0 +1,350 @@
+
+The current implementation of Addons has some flaws that make the usage and maintenance of Addons cumbersome for cluster administrators:
+
+- To add “Custom Addons” the cluster-admin needs to bundle manifests and templates for these Addons in a Docker image that is based on the official addon-manifest docker image by Kubermatic.
```
Besides that in most of the cases user needs to create AddonConfig CRD for each custom/accessible addon, so form, logo and description will be visible in the dashboard.

maciaszczykm replied:
See: https://github.com/kubermatic/kubermatic/issues/5499

SimonTheLeg replied:
Good point, I have referenced the PR in the proposal

maciaszczykm commented on docs/proposals/improve-addons.md:
```diff
@@ -0,0 +1,350 @@
+- Decouple custom addons from the KKP lifecycle
+- Handle dependency **inside** an Addon (e.g. CRD need to be created before CRs, namespace needs to be created before other objects)
+- Addons are correctly reconciled on changes. [kubermatic/kubermatic#6035](https://github.com/kubermatic/kubermatic/issues/6035)
+- Be able to choose which addons should be installed during the cluster creation or in the cluster template
```
Linking: https://github.com/kubermatic/kubermatic/issues/6000

SimonTheLeg replied:
I just realized from re-reading this bullet, that choosing suitable addons before cluster-creation is something that will be possible with the components in this proposal, but is not included as a design decision

cc @vgramer: I think we need to think about whether we want to put this into this proposal or in a separate one/follow up User Story

SimonTheLeg replied:
@maciaszczykm

very good point. We actually realized that there was required functionality missing in the architecture. You can find it in the AddonController section as a separate bulletpoint.

Additionally I think we also need to update #6000 later on, as choosing addons is a bit more complicated due to the fact that the compatible addons need to be computed based on the chosen cluster.

maciaszczykm commented on docs/proposals/improve-addons.md:
```diff
@@ -0,0 +1,350 @@
+     - prometheus:
+            description: "The Prometheus Node Exporter exposes a wide variety of hardware- and kernel-related metrics."
+            logoFormat: "svg+xml"
+            logo: # logo must be base64 encoded.
```
Size of it may be problematic if we will store the logos together. In the @kubermatic/sig-api @kubermatic/sig-ui we were considering a change here, perhaps we should aim to host these images somewhere.

SimonTheLeg replied:
good point. I have added logos as a non-goal for now, so we can move forward with the topic. Good news is: With the current architecture, we can add logos later on without any problems; be it base64 encoded or hosted somewhere else

maciaszczykm commented on docs/proposals/improve-addons.md:
```diff
@@ -0,0 +1,350 @@
+- Handle dependency **inside** an Addon (e.g. CRD need to be created before CRs, namespace needs to be created before other objects)
+- Addons are correctly reconciled on changes. [kubermatic/kubermatic#6035](https://github.com/kubermatic/kubermatic/issues/6035)
+- Be able to choose which addons should be installed during the cluster creation or in the cluster template
+- Be compatible with private registries in preparation for future air-gapped environment stories. The current Policy regarding the air-gapped environment is to assume a private trusted registry is available. Addon Source (manifest of the workload to deployed) and workload images of the addon should be  configurable to point on a private registry
```
Just an idea, but if we are improving the addons then perhaps we can solve issue like https://github.com/kubermatic/dashboard/issues/3666 too. So make k8s/dashboard an addon that will be always deployed but the version used will differ based on the user cluster version.

SimonTheLeg replied:
Yes this would definitely be possible. You could even use constraints to ensure only a suitable dashboard is being shown to users. 

While it's not directly a goal of this proposal, I am going to add it to the "Non Goals section". This ensures that we have the topic ready on our roadmap :)

SimonTheLeg committed:
Include Suggestions:

- dashboard rework in non-goals
- handling logos in non-goals
- fix yaml indents
- fix typos

Signed-off-by: Simon Bein <simontheleg@gmail.com>

SimonTheLeg committed:
Include Suggestions:

    - add more constraints
    - protect coreaddons using webhooks
    - injecting template data into helm-charts
    - allowing for addon selection in the cluster creation wizard
    - handle dependencies in go-template method

Signed-off-by: Simon Bein <simontheleg@gmail.com>

embik reviewed (approved):
/lgtm

I'm however not sure I'm the right person to "greenlight" this, so it's a good thing that others are taking a look too.

themue reviewed (approved):
/approve
/lgtm

vgramer reviewed (approved):
/lgtm
